domain/src/main/java/com/github/dodobest/domain/Expression.kt

```diff
@@ -0,0 +1,132 @@
+package com.github.dodobest.domain
+
+class Expression {
+    fun evaluate(input: String): List<String> {
+        val inputString: String = eraseBlank(input)
+
+        require(inputString != "") { "빈 공백문자를 입력했습니다." }
+
+        return extractNumAndSignAll(inputString)
+    }
+
+    private fun eraseBlank(inputString: String): String {
+        return inputString.replace(" ", "")
+    }
+
+
+    private fun extractNumAndSignAll(inputString: String): List<String> {
+        val numAndSignArray: MutableList<String> = MutableList(0) { "" }
+        val firstIndexOfNum: Array<Int> = arrayOf(-1)
+
+        for (idx in inputString.indices) {
+            numAndSignArray.addAll(splitNumAndSignOnce(inputString, idx, firstIndexOfNum))
+        }
+        numAndSignArray.add(inputString.slice(IntRange(firstIndexOfNum[0], inputString.length-1)))
+
+        return numAndSignArray
+    }
+
+    private fun splitNumAndSignOnce(inputString: String, charIndex: Int, firstIndexOfNum: Array<Int>): List<String> {
+        if (isNum(inputString[charIndex])) {
+            updateFirstIndexOfNum(inputString, charIndex, firstIndexOfNum)
+            return MutableList(0) { "" }
+        }
+
+        return checkArithmeticOperation(inputString, charIndex, firstIndexOfNum)
+    }
+
+    private fun updateFirstIndexOfNum(inputString: String, charIndex: Int, firstIndexOfNum: Array<Int>){
+        if (firstIndexOfNum[0] != -1) {
+            return
+        }
+
+        require(!checkNumStartWithZeroAndNotExactZero(inputString, charIndex)) { "0으로 시작하는 숫자는 지원하지 않습니다." }
+
+        firstIndexOfNum[0] = charIndex
+    }
+
+    fun isNum(charVal: Char): Boolean {
+        if (charVal.code in '0'.code..'9'.code ) {
+            return true
+        }
+        return false
+    }
+
+    private fun checkNumStartWithZeroAndNotExactZero(inputString: String, charIndex: Int): Boolean {
+        return inputString[charIndex] == '0' && charIndex < inputString.length - 1 && isNum(inputString[charIndex + 1])
+    }
+
+    private fun checkArithmeticOperation(inputString: String, charIndex: Int, firstIndexOfNum: Array<Int>): List<String> {
+        val numAndSignArray: MutableList<String> = MutableList(0) { "" }
+
+        checkInputIsNotCorrect(inputString, charIndex)
+
+        if (isNegativeSignNotMinusSign(inputString[charIndex], firstIndexOfNum)) {
+            firstIndexOfNum[0] = charIndex
+            return MutableList(0) { "" }
+        }
+
+        if (isPositiveSign(inputString[charIndex], firstIndexOfNum)) {
+            return MutableList(0) { "" }
+        }
+
+        numAndSignArray.add(inputString.slice(IntRange(firstIndexOfNum[0], charIndex-1)))
+        numAndSignArray.add(inputString[charIndex].toString())
+
+        firstIndexOfNum[0] = -1
+
+        return numAndSignArray
```

Comment:
```suggestion
        return numAndSignArray.toList()
```
가급적이면 방어적 복사를 해서 리턴해주는 습관을 들이시면 좋습니다.
방어적 복사를 하게 되면 리스트를 새로 만들기 때문에, 원본이 외부에서 수정될 가능성이 없어집니다.
(이하 관련 내용 동일)